# Carmencita / CARM_VIS DATABASE LOADING AND INSPECTION
Identifying CARM_VIS objects in Carmencita database, and checking if TESS light curve (LC) exist for each object
(With pandas)

## Modules and configuration

### Modules

In [1]:
# Modules import:
import numpy as np
import pandas as pd
from os import listdir
from os.path import isfile, join

### Configuration

In [2]:
# Configuration:
# Files:
CARMENES_DBFILE = "../data/carmencita.102.csv"
CARMENES_MODDBFILE = CARMENES_DBFILE[:-4] + "_mod_CARM_VIS.csv"
CARMENES_STATS_OUT = CARMENES_DBFILE[:-4] + "_stats_CARM_VIS.csv"
CARMENES_NAN_OUT = CARMENES_DBFILE[:-4] + "_nan_CARM_VIS.csv"

GTO_FOLDER = "../data/CARM_VIS_RVs/"
GTO_FOLDER_LC = "../data/CARMENES_GTO_TESS_lc/lightcurves"
LATE_GTO_FOLDER_LC = "../data/CARMENES_GTO_TESS_lc/CARMENES_late"

## Functions

### Get a range from two columns (nominal and error)

### Locate TESS light curve

In [3]:
def locate_lc(obj_name, folder):
    try:
        files_in_folder = "/".join([folder, obj_name])
        #print(files_in_folder)
        files_in_folder = listdir(files_in_folder)
        #print(files_in_folder)
        if len(files_in_folder) > 1:
            filename = "<ambiguous>"
        else:
            filename = "/".join([folder, obj_name, files_in_folder[0]])
    except:
        filename = None
    return filename

In [4]:
# TEST:
locate_lc("J00051+457", GTO_FOLDER_LC)

'../data/CARMENES_GTO_TESS_lc/lightcurves/J00051+457/J00051+457_TESS_TIC439946126_S17.fits'

In [5]:
# TEST:
a = locate_lc("unknown_name", GTO_FOLDER_LC)
a is None

True

## Data loading and combining

### Load CARMENES database file

In [6]:
carmenes = pd.read_csv(CARMENES_DBFILE, sep=',', decimal='.')
carmenes.head()

Karmn           Name Comp    GJ  RA_J2016_deg  DE_J2016_deg  \
0  J23598+477     LP 149-014    -  4385    359.962056     47.761478   
1  J23590+208      G 129-051    -   NaN    359.753066     20.860336   
2  J23587+467  BD+45 4378 AB   AB   913    359.685432     46.729152   
3  J23585+076      Wolf 1051   AB  4383    359.636420      7.656947   
4  J23582-174     LP 764-040   AB   NaN    359.558095    -17.409041   

      RA_J2000     DE_J2000  l_J2016_deg  b_J2016_deg  ... LoRes_spectrum  \
0  23:59:49.51  +47:45:44.5   114.027674   -14.216353  ...            NaN   
1  23:59:00.42  +20:51:39.0   106.785616   -40.361996  ...          CAFOS   
2  23:58:43.49  +46:43:44.7   113.616244   -15.187224  ...            NaN   
3  23:58:32.65  +07:39:30.1   100.839686   -52.931049  ...            NaN   
4  23:58:13.69  -17:24:32.9    68.527865   -74.336758  ...            NaN   

  HiRes_spectrum  LoRes_imaging       HiRes_imaging   Class  Survey  \
0            NaN            NaN                 NaN   Gamma     NaN   
1            NaN            NaN                 NaN   Delta     NaN   
2            NaN            NaN  (FastCam) [Bale07]   Delta     NaN   
3          FEROS            NaN           (FastCam)  Delta+     GTO   
4            NaN            NaN    [Dae07 AstraLux]   Delta     NaN   

                                 Notes            Gaia_id_1          2MASS_id  \
0                                  NaN  1927517417630958208  23594940+4745447   
1                                  NaN  2846620421604285568  23590042+2051387   
2  # WDS BAG34 consistent wtih Hei415.  1927184643563033856  23584342+4643452   
3                                  NaN  2746709861572970624  23583264+0739304   
4                     #20-150Ma Shk09.  2414454421367861376  23581366-1724338   

       TIC_id  
0   177697529  
1  2053997903  
2  2041238829  
3   408486498  
4  2052312661  

[5 rows x 175 columns]

In [7]:
print(carmenes.columns.tolist())

['Karmn', 'Name', 'Comp', 'GJ', 'RA_J2016_deg', 'DE_J2016_deg', 'RA_J2000', 'DE_J2000', 'l_J2016_deg', 'b_J2016_deg', 'Ref01', 'SpT', 'SpTnum', 'Ref02', 'Teff_K', 'eTeff_K', 'logg', 'elogg', '[Fe/H]', 'e[Fe/H]', 'Ref03', 'L_Lsol', 'eL_Lsol', 'Ref04', 'R_Rsol', 'eR_Rsol', 'Ref05', 'M_Msol', 'eM_Msol', 'Ref06', 'muRA_masa-1', 'emuRA_masa-1', 'muDE_masa-1', 'emuDE_masa-1', 'Ref07', 'pi_mas', 'epi_mas', 'Ref08', 'd_pc', 'ed_pc', 'Ref09', 'Vr_kms-1', 'eVr_kms-1', 'Ref10', 'ruwe', 'Ref11', 'U_kms-1', 'eU_kms-1', 'V_kms-1', 'eV_kms-1', 'W_kms-1', 'eW_kms-1', 'Ref12', 'sa_m/s/a', 'esa_m/s/a', 'Ref13', 'SKG', 'Ref14', 'SKG_lit', 'Ref14_lit', 'Pop', 'Ref15', 'vsini_flag', 'vsini_kms-1', 'evsini_kms-1', 'Ref16', 'P_d', 'eP_d', 'Ref17', 'pEWHalpha_A', 'epEWHalpha_A', 'Ref18', 'log(LHalpha/Lbol)', 'elog(LHalpha/Lbol)', 'Ref19', '1RXS', 'CRT_s-1', 'eCRT_s-1', 'HR1', 'eHR1', 'HR2', 'eHR2', 'Flux_X_E-13_ergcm-2s-1', 'eFlux_X_E-13_ergcm-2s-1', 'LX/LJ', 'eLX/LJ', 'Ref20', 'Activity', 'Ref21', 'FUV_mag',

### Basic statistics

In [8]:
carmenes_stats = carmenes.describe()
carmenes_stats

RA_J2016_deg  DE_J2016_deg  l_J2016_deg  b_J2016_deg       SpTnum  \
count   2212.000000   2212.000000  2211.000000  2211.000000  2212.000000   
mean     174.478191     25.033073   141.530909    10.808282     2.804566   
std      105.175380     25.873700    77.624600    38.548382     1.571185   
min        0.303710    -32.437037     0.297124   -85.781428    -2.000000   
25%       79.220264      5.438041    80.737176   -18.876011     1.500000   
50%      170.592594     22.680032   137.796626    12.442436     3.000000   
75%      265.664080     44.133004   191.232163    42.360953     4.000000   
max      359.962056     88.405627   359.808867    89.717950     9.500000   

            Teff_K      eTeff_K         logg        elogg       [Fe/H]  ...  \
count  1657.000000  1657.000000  1652.000000  1652.000000  1652.000000  ...   
mean   3422.496077    45.379602     4.978553     0.410533    -0.030448  ...   
std     283.857097    11.216744     0.355109     0.160944     0.092074  ...   
min    2400.000000     9.000000     4.000000     0.030000    -0.620000  ...   
25%    3200.000000    50.000000     4.687500     0.500000     0.000000  ...   
50%    3400.000000    50.000000     5.000000     0.500000     0.000000  ...   
75%    3600.000000    50.000000     5.140000     0.500000     0.000000  ...   
max    4500.000000    97.000000     5.500000     0.500000     0.420000  ...   

           eW1_mag       W2_mag      eW2_mag       W3_mag      eW3_mag  \
count  2193.000000  2194.000000  2194.000000  2194.000000  2194.000000   
mean      0.043555     7.536307     0.025103     7.454438     0.019887   
std       0.045817     1.038403     0.021990     0.962119     0.005868   
min       0.011000     2.360000     0.010000     3.045000     0.010000   
25%       0.023000     7.061250     0.020000     6.988250     0.016000   
50%       0.027000     7.678500     0.020000     7.576500     0.018000   
75%       0.039000     8.204250     0.021000     8.082750     0.021000   
max       0.524000    11.316000     0.429000    10.313000     0.091000   

            W4_mag      eW4_mag    PlanetNum     Gaia_id_1        TIC_id  
count  2194.000000  2115.000000  2206.000000  2.212000e+03  2.212000e+03  
mean      7.309170     0.146526     0.036718  2.313430e+18  3.593196e+08  
std       0.934874     0.097991     0.284143  1.567342e+18  3.709552e+08  
min       2.934000     0.015000     0.000000  3.179036e+15  7.073460e+05  
25%       6.852250     0.077000     0.000000  9.819256e+17  1.442746e+08  
50%       7.444000     0.121000     0.000000  2.104625e+18  2.859379e+08  
75%       7.920750     0.188000     0.000000  3.409167e+18  4.173732e+08  
max       9.495000     0.540000     7.000000  6.916735e+18  2.054010e+09  

[8 rows x 99 columns]

In [9]:
carmenes_nan = carmenes.isnull().sum()
carmenes_nan

Karmn              0
Name               0
Comp               0
GJ               914
RA_J2016_deg       0
                ... 
Survey          1811
Notes           1868
Gaia_id_1          0
2MASS_id           0
TIC_id             0
Length: 175, dtype: int64

### Add ranges for `Teff_K` and `logg`variables

In [10]:
carmenes['Teff_min_K'] = carmenes['Teff_K'] - carmenes['eTeff_K']
carmenes['Teff_max_K'] = carmenes['Teff_K'] + carmenes['eTeff_K']
carmenes['logg_min'] = carmenes['logg'] - carmenes['elogg']
carmenes['logg_max'] = carmenes['logg'] + carmenes['elogg']


### Retrieve GTO object names

In [16]:
dir_items = listdir(GTO_FOLDER)
gto_names = []
for f in dir_items:
    try:
        if isfile(join(GTO_FOLDER, f)) and f[-7:] == "avc.dat":
            # Incorporate object name:
            gto_names.append(f[:f.find('.')])
        else:
            pass
    except:
        pass
print("%d GTO objects found." %len(gto_names))
print("10 first GTO objects: %s" %gto_names[:10])

390 GTO objects found.
10 first GTO objects: ['J00051+457', 'J00067-075', 'J00162+198E', 'J00162+198W', 'J00183+440', 'J00184+440', 'J00286-066', 'J00389+306', 'J00403+612', 'J00570+450']


### Identify GTO objects in CARMENES database and locate RV files

In [17]:
carmenes['is_GTO'] = False
carmenes.loc[carmenes['Karmn'].isin(gto_names), 'is_GTO'] = True
print("GTO objects identified in CARMENES database: %d" %carmenes[carmenes['is_GTO']].shape[0])

GTO objects identified in CARMENES database: 390


In [18]:
carmenes['rv_file'] = None
carmenes.loc[carmenes['is_GTO'] == True, 'rv_file'] = GTO_FOLDER + \
    carmenes.loc[carmenes['is_GTO'] == True, 'Karmn'] + ".dat"

### Locate TESS light curve files

We first try with CARMENES "late" folder, with only 16 files.

In [19]:
carmenes['lc_file'] = carmenes['Karmn'].map(lambda x: locate_lc(x, LATE_GTO_FOLDER_LC))
carmenes.head()

Karmn           Name Comp    GJ  RA_J2016_deg  DE_J2016_deg  \
0  J23598+477     LP 149-014    -  4385    359.962056     47.761478   
1  J23590+208      G 129-051    -   NaN    359.753066     20.860336   
2  J23587+467  BD+45 4378 AB   AB   913    359.685432     46.729152   
3  J23585+076      Wolf 1051   AB  4383    359.636420      7.656947   
4  J23582-174     LP 764-040   AB   NaN    359.558095    -17.409041   

      RA_J2000     DE_J2000  l_J2016_deg  b_J2016_deg  ...  \
0  23:59:49.51  +47:45:44.5   114.027674   -14.216353  ...   
1  23:59:00.42  +20:51:39.0   106.785616   -40.361996  ...   
2  23:58:43.49  +46:43:44.7   113.616244   -15.187224  ...   
3  23:58:32.65  +07:39:30.1   100.839686   -52.931049  ...   
4  23:58:13.69  -17:24:32.9    68.527865   -74.336758  ...   

             Gaia_id_1          2MASS_id      TIC_id Teff_min_K  Teff_max_K  \
0  1927517417630958208  23594940+4745447   177697529     2950.0      3050.0   
1  2846620421604285568  23590042+2051387  2053997903        NaN         NaN   
2  1927184643563033856  23584342+4643452  2041238829        NaN         NaN   
3  2746709861572970624  23583264+0739304   408486498     3476.0      3516.0   
4  2414454421367861376  23581366-1724338  2052312661        NaN         NaN   

   logg_min  logg_max  is_GTO                              rv_file  lc_file  
0      4.50      5.50   False                                 None     None  
1       NaN       NaN   False                                 None     None  
2       NaN       NaN   False                                 None     None  
3      4.89      5.05    True  ../data/CARM_VIS_RVs/J23585+076.dat     None  
4       NaN       NaN   False                                 None     None  

[5 rows x 182 columns]

In [20]:
carmenes[(carmenes['is_GTO'] == True) & (carmenes['lc_file'].isna())].shape

(379, 182)

In [21]:
carmenes['lc_file'] = carmenes['Karmn'].map(lambda x: locate_lc(x, GTO_FOLDER_LC))
carmenes.head()

Karmn           Name Comp    GJ  RA_J2016_deg  DE_J2016_deg  \
0  J23598+477     LP 149-014    -  4385    359.962056     47.761478   
1  J23590+208      G 129-051    -   NaN    359.753066     20.860336   
2  J23587+467  BD+45 4378 AB   AB   913    359.685432     46.729152   
3  J23585+076      Wolf 1051   AB  4383    359.636420      7.656947   
4  J23582-174     LP 764-040   AB   NaN    359.558095    -17.409041   

      RA_J2000     DE_J2000  l_J2016_deg  b_J2016_deg  ...  \
0  23:59:49.51  +47:45:44.5   114.027674   -14.216353  ...   
1  23:59:00.42  +20:51:39.0   106.785616   -40.361996  ...   
2  23:58:43.49  +46:43:44.7   113.616244   -15.187224  ...   
3  23:58:32.65  +07:39:30.1   100.839686   -52.931049  ...   
4  23:58:13.69  -17:24:32.9    68.527865   -74.336758  ...   

             Gaia_id_1          2MASS_id      TIC_id Teff_min_K  Teff_max_K  \
0  1927517417630958208  23594940+4745447   177697529     2950.0      3050.0   
1  2846620421604285568  23590042+2051387  2053997903        NaN         NaN   
2  1927184643563033856  23584342+4643452  2041238829        NaN         NaN   
3  2746709861572970624  23583264+0739304   408486498     3476.0      3516.0   
4  2414454421367861376  23581366-1724338  2052312661        NaN         NaN   

   logg_min  logg_max  is_GTO                              rv_file  lc_file  
0      4.50      5.50   False                                 None     None  
1       NaN       NaN   False                                 None     None  
2       NaN       NaN   False                                 None     None  
3      4.89      5.05    True  ../data/CARM_VIS_RVs/J23585+076.dat     None  
4       NaN       NaN   False                                 None     None  

[5 rows x 182 columns]

In [22]:
carmenes[(carmenes['is_GTO'] == True) & (carmenes['lc_file'].isna())].shape

(123, 182)

In [23]:
carmenes.tail()

Karmn                     Name Comp   GJ  RA_J2016_deg  \
2207   J00051+457                     GJ 2    C    2      1.300921   
2208   J00033+046              StKM 1-2199    -  NaN      0.829061   
2209   J00026+383  2MASS J00024011+3821453    -  NaN      0.666656   
2210  J00012+139S              BD+13 5195B    B  NaN      0.303710   
2211  J00012+139N               BD+13 5195    -  NaN      0.305059   

      DE_J2016_deg     RA_J2000     DE_J2000  l_J2016_deg  b_J2016_deg  ...  \
2207     45.785894  00:05:10.89  +45:47:11.6   114.559789   -16.332933  ...   
2208      4.686556  00:03:19.00  +04:41:13.0   101.048765   -56.127008  ...   
2209     38.362250  00:02:40.09  +38:21:44.5   112.526509   -23.522080  ...   
2210     13.972782  00:01:12.86  +13:58:19.6   104.868071   -47.129826  ...   
2211     13.975743  00:01:13.20  +13:58:30.3   104.871154   -47.127359  ...   

                Gaia_id_1          2MASS_id     TIC_id Teff_min_K  Teff_max_K  \
2207   386655019234959872  00051079+4547116  439946126     3757.0      3789.0   
2208  2741901697224268928  00031900+0441128  257503361     3650.0      3750.0   
2209  2880923878720791680  00024011+3821453  612030125        NaN         NaN   
2210  2767384425386994048  00011286+1358196  405230222     3950.0      4050.0   
2211  2767387380324493184  00011318+1358302  405230221        NaN         NaN   

      logg_min  logg_max  is_GTO                              rv_file  \
2207      4.99      5.15    True  ../data/CARM_VIS_RVs/J00051+457.dat   
2208      4.50      5.50   False                                 None   
2209       NaN       NaN   False                                 None   
2210      4.50      5.50   False                                 None   
2211       NaN       NaN   False                                 None   

                                                lc_file  
2207  ../data/CARMENES_GTO_TESS_lc/lightcurves/J0005...  
2208                                               None  
2209                                               None  
2210                                               None  
2211                                               None  

[5 rows x 182 columns]

Now we check the numbers:

Number of records with RV file (should be all GTO objects):

In [24]:
carmenes[~carmenes['rv_file'].isna()].shape[0]

390

Number of records with TESS LC file:

In [25]:
carmenes[~carmenes['lc_file'].isna()].shape[0]

267

Number of GTO records without RV file (shouldn't be any):

In [26]:
carmenes[(carmenes['is_GTO'] == True) & (carmenes['rv_file'].isna())].shape[0]

0

Number of GTO records without TESS LC file:

In [27]:
carmenes[(carmenes['is_GTO'] == True) & (carmenes['lc_file'].isna())].shape[0]

123

Number of records with the light curve under "late" folder:

In [28]:
carmenes[(carmenes['is_GTO'] == True) & (carmenes['lc_file'].str.contains("CARMENES_late"))].shape[0]

0

**OBSERVATION:** we have 123 GTO objects for which we do not have TESS light curve. This could be normal, as some GTO objects are out of the TESS field.

### Add RV file and LC file flags

In [29]:
carmenes['has_rv'] = carmenes['rv_file'].map(lambda x: False if x is None else True)
carmenes['has_lc'] = carmenes['lc_file'].map(lambda x: False if x is None else True)
carmenes.head()

Karmn           Name Comp    GJ  RA_J2016_deg  DE_J2016_deg  \
0  J23598+477     LP 149-014    -  4385    359.962056     47.761478   
1  J23590+208      G 129-051    -   NaN    359.753066     20.860336   
2  J23587+467  BD+45 4378 AB   AB   913    359.685432     46.729152   
3  J23585+076      Wolf 1051   AB  4383    359.636420      7.656947   
4  J23582-174     LP 764-040   AB   NaN    359.558095    -17.409041   

      RA_J2000     DE_J2000  l_J2016_deg  b_J2016_deg  ...      TIC_id  \
0  23:59:49.51  +47:45:44.5   114.027674   -14.216353  ...   177697529   
1  23:59:00.42  +20:51:39.0   106.785616   -40.361996  ...  2053997903   
2  23:58:43.49  +46:43:44.7   113.616244   -15.187224  ...  2041238829   
3  23:58:32.65  +07:39:30.1   100.839686   -52.931049  ...   408486498   
4  23:58:13.69  -17:24:32.9    68.527865   -74.336758  ...  2052312661   

  Teff_min_K  Teff_max_K logg_min  logg_max  is_GTO  \
0     2950.0      3050.0     4.50      5.50   False   
1        NaN         NaN      NaN       NaN   False   
2        NaN         NaN      NaN       NaN   False   
3     3476.0      3516.0     4.89      5.05    True   
4        NaN         NaN      NaN       NaN   False   

                               rv_file  lc_file  has_rv  has_lc  
0                                 None     None   False   False  
1                                 None     None   False   False  
2                                 None     None   False   False  
3  ../data/CARM_VIS_RVs/J23585+076.dat     None    True   False  
4                                 None     None   False   False  

[5 rows x 184 columns]

### Save the modified information

In [30]:
carmenes.shape

(2212, 184)

In [31]:
carmenes_stats.to_csv(CARMENES_STATS_OUT, sep=',', decimal='.')

In [32]:
carmenes_nan.to_csv(CARMENES_NAN_OUT, sep=',', decimal='.')

In [33]:
carmenes.to_csv(CARMENES_MODDBFILE, sep=',', decimal='.', index=False)